File: data_mining/auto_encoder/JavaScript/auto_encoder.js
// Fonction d'activation tanh
function tanh(x) {
    return Math.tanh(x);
  }
  
  // Dérivée de la fonction d'activation tanh
  function tanhPrime(x) {
    return 1 - Math.pow(tanh(x), 2);
  }
  
  // Initialisation des poids de manière aléatoire
  function initializeWeights(inputSize, hiddenSize, outputSize) {
    // Poids entre l'entrée et la couche cachée
    let weightsInputHidden = new Array(inputSize).fill().map(() =>
      new Array(hiddenSize).fill().map(() => Math.random() - 0.5)
    );
    // Poids entre la couche cachée et la sortie
    let weightsHiddenOutput = new Array(hiddenSize).fill().map(() =>
      new Array(outputSize).fill().map(() => Math.random() - 0.5)
    );
  
    return { weightsInputHidden, weightsHiddenOutput };
  }
  
  // Propagation avant pour obtenir les sorties cachées et les sorties finales
  function forwardPass(input, weights) {
    let hidden = weights.weightsInputHidden.map(h =>
      tanh(input.reduce((sum, inp, idx) => sum + inp * h[idx], 0))
    );
  
    let output = weights.weightsHiddenOutput.map(o =>
      tanh(hidden.reduce((sum, hid, idx) => sum + hid * o[idx], 0))
    );
  
    return { hidden, output };
  }
  
  // Calcul de l'erreur et de la dérivée
  function computeErrorAndGradient(target, output, hidden, weights) {
    let outputError = output.map((o, idx) => target[idx] - o);
    let outputGradient = outputError.map((e, idx) => e * tanhPrime(output[idx]));
  
    let hiddenError = weights.weightsHiddenOutput.map(h =>
      outputGradient.reduce((sum, og, ogIdx) => sum + og * h[ogIdx], 0)
    );
    let hiddenGradient = hiddenError.map((e, idx) => e * tanhPrime(hidden[idx]));
  
    return { outputError, outputGradient, hiddenError, hiddenGradient };
  }
  
  // Mise à jour des poids
  function updateWeights(weights, gradients, learningRate) {
    // Mise à jour des poids de la couche cachée vers la sortie
    for (let i = 0; i < weights.weightsHiddenOutput.length; i++) {
      for (let j = 0; j < weights.weightsHiddenOutput[i].length; j++) {
        weights.weightsHiddenOutput[i][j] += gradients.outputGradient[j] * learningRate;
      }
    }
  
    // Mise à jour des poids de l'entrée vers la couche cachée
    for (let i = 0; i < weights.weightsInputHidden.length; i++) {
      for (let j = 0; j < weights.weightsInputHidden[i].length; j++) {
        weights.weightsInputHidden[i][j] += gradients.hiddenGradient[j] * learningRate;
      }
    }
  
    return weights;
  }
  
  // Entraînement de l'auto-encodeur
  function trainAutoEncoder(input, target, inputSize, hiddenSize, outputSize, epochs, learningRate) {
    let weights = initializeWeights(inputSize, hiddenSize, outputSize);
  
    for (let epoch = 0; epoch < epochs; epoch++) {
      input.forEach((inp, index) => {
        // Propagation avant pour un élément d'entrée unique
        let { hidden, output } = forwardPass(inp, weights);
  
        // Cible correspondant à l'entrée courante
        let currentTarget = target[index];
  
        // Calcul de l'erreur et du gradient pour l'élément courant
        let gradients = computeErrorAndGradient(currentTarget, output, hidden, weights);
  
        // Mise à jour des poids
        weights = updateWeights(weights, gradients, learningRate);
      });
    }
  
    return weights;
  }
  
  // Données d'entraînement
  let input = [[0, 1], [1, 0], [1, 1], [0, 0]];
  let target = input; // Dans un auto-encodeur, la cible est la même que l'entrée
  let inputSize = 2;
  let hiddenSize = 2;
  let outputSize = 2;
  let epochs = 10000;
  let learningRate = 0.01;
  
  // Entraînement
  let trainedWeights = trainAutoEncoder(input, target, inputSize, hiddenSize, outputSize, epochs, learningRate);
  
  // Affichage des résultats après entraînement
  input.forEach(i => {
    let { output } = forwardPass(i, trainedWeights);
    console.log(`Input: ${i} Output: ${output.map(o => o.toFixed(2))}`); // Utilisation de toFixed(2) pour limiter le nombre de décimales
  });
  

File: data_mining/auto_encoder/Python/without_dependence/auto_encoder.py
# flake8: noqa
import random
import math

# Fonction d'activation tanh
def tanh(x):
    return math.tanh(x)

# Dérivée de la fonction d'activation tanh
def tanh_prime(x):
    return 1 - math.tanh(x)**2

# Initialisation des poids de manière aléatoire
def initialize_weights(input_size, hidden_size, output_size):
    weights_input_hidden = [[random.random() - 0.5 for _ in range(hidden_size)] for _ in range(input_size)]
    weights_hidden_output = [[random.random() - 0.5 for _ in range(output_size)] for _ in range(hidden_size)]
    return weights_input_hidden, weights_hidden_output

# Propagation avant pour obtenir les sorties cachées et les sorties finales
def forward_pass(input, weights_input_hidden, weights_hidden_output):
    hidden = [tanh(sum(i * w for i, w in zip(input, weights))) for weights in weights_input_hidden]
    output = [tanh(sum(h * w for h, w in zip(hidden, weights))) for weights in weights_hidden_output]
    return hidden, output

# Calcul de l'erreur et de la dérivée
def compute_error_and_gradient(target, output, hidden, weights_hidden_output):
    output_error = [t - o for t, o in zip(target, output)]
    output_gradient = [e * tanh_prime(o) for e, o in zip(output_error, output)]

    hidden_error = [sum(og * w for og, w in zip(output_gradient, col)) for col in zip(*weights_hidden_output)]
    hidden_gradient = [e * tanh_prime(h) for e, h in zip(hidden_error, hidden)]

    return output_error, output_gradient, hidden_error, hidden_gradient

# Mise à jour des poids
def update_weights(weights_input_hidden, weights_hidden_output, input, hidden, output_gradient, hidden_gradient, learning_rate):
    for i, (wg, h) in enumerate(zip(output_gradient, hidden)):
        for j in range(len(weights_hidden_output[i])):
            weights_hidden_output[i][j] += wg * h * learning_rate

    for i, (hg, inp) in enumerate(zip(hidden_gradient, input)):
        for j in range(len(weights_input_hidden[i])):
            weights_input_hidden[i][j] += hg * inp * learning_rate

    return weights_input_hidden, weights_hidden_output

# Entraînement de l'auto-encodeur
def train_auto_encoder(input, target, input_size, hidden_size, output_size, epochs, learning_rate):
    weights_input_hidden, weights_hidden_output = initialize_weights(input_size, hidden_size, output_size)

    for epoch in range(epochs):
        for inp, tar in zip(input, target):
            hidden, output = forward_pass(inp, weights_input_hidden, weights_hidden_output)
            output_error, output_gradient, hidden_error, hidden_gradient = compute_error_and_gradient(tar, output, hidden, weights_hidden_output)
            weights_input_hidden, weights_hidden_output = update_weights(weights_input_hidden, weights_hidden_output, inp, hidden, output_gradient, hidden_gradient, learning_rate)

    return weights_input_hidden, weights_hidden_output

# Données d'entraînement
input = [[0, 1], [1, 0], [1, 1], [0, 0]]
target = input  # Dans un auto-encodeur, la cible est la même que l'entrée
input_size = 2
hidden_size = 2
output_size = 2
epochs = 10000
learning_rate = 0.01

# Entraînement
weights_input_hidden, weights_hidden_output = train_auto_encoder(input, target, input_size, hidden_size, output_size, epochs, learning_rate)

# Affichage des résultats après entraînement
for i in input:
    _, output = forward_pass(i, weights_input_hidden, weights_hidden_output)
    print(f"Input: {i} Output: {[round(o, 2) for o in output]}")
